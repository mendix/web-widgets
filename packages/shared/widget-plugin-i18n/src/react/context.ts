import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
