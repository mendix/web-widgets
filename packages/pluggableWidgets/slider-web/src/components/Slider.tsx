import { createElement, CSSProperties, forwardRef, RefObject } from "react";
import RcSlider, { SliderProps as RcSliderProps } from "rc-slider";
import classNames from "classnames";

export interface SliderProps extends RcSliderProps {
    classNameSlider?: string;
    rootStyle?: CSSProperties;
}

export const Slider = forwardRef(
    (
        { className, classNameSlider, rootStyle, ...rcSliderProps }: SliderProps,
        sliderRef: RefObject<HTMLDivElement>
    ): JSX.Element => {
        return (
            <div
                ref={sliderRef}
                style={rootStyle}
                className={classNames(
                    "widget-slider",
                    {
                        "widget-slider-vertical": rcSliderProps.vertical
                    },
                    className
                )}
            >
                <RcSlider className={classNameSlider} {...rcSliderProps} />
            </div>
        );
    }
);
