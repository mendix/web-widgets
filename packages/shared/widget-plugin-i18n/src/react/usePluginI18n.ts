import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
