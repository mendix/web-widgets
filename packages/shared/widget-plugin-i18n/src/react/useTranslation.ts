import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
