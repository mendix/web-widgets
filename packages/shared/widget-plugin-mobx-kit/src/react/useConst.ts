import { useRef } from "react";

export function useConst<T>(fn: T): T extends (...args: never) => infer R ? R : T {
    return (useRef<T | null>(null).current ??= typeof fn === "function" ? fn() : fn);
}
