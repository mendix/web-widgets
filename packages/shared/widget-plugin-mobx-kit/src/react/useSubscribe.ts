import { IReactionOptions, reaction } from "mobx";
import { useEffect, useState } from "react";

export function useSubscribe<Store extends object, T, FireImmediately extends boolean = false>(
    init: Store | (() => Store),
    selector: (store: Store) => T,
    options?: IReactionOptions<T, FireImmediately>
): [T, Store] {
    const [store] = useState(() => (typeof init === "function" ? init() : init));
    const [value, setValue] = useState(() => selector(store));

    function setup(): () => void {
        return reaction(() => selector(store), setValue, options);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(setup, []);

    return [value, store];
}
