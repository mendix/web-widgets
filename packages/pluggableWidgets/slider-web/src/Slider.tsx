import { createElement, ReactNode } from "react";
import { SliderContainerProps } from "../typings/SliderProps";
import { getMinValue } from "./utils/getMinValue";
import { getMaxValue } from "./utils/getMaxValue";
import { getStepValue } from "./utils/getStepValue";
import { useMarks } from "./utils/useMarks";
import { createHandleGenerator } from "./utils/createHandleGenerator";
import { useOnChangeDebounced } from "./utils/useOnChangeDebounced";
import { Slider as SliderComponent } from "./components/Slider";
import "./ui/Slider.scss";
import { isVertical } from "./utils/isVertical";
import { getStyleProp } from "./utils/getStyleProp";
import { useScheduleUpdateOnce } from "@mendix/widget-plugin-hooks/useScheduleUpdateOnce";
import { ValueStatus } from "mendix";

export default function Slider(props: SliderContainerProps): ReactNode {
    const {
        valueAttribute,
        orientation,
        heightUnit,
        height,
        minValueType,
        minAttribute,
        expressionMinimumValue,
        staticMinimumValue,
        stepSizeType,
        stepValue,
        stepAttribute,
        expressionStepSize,
        maxValueType,
        staticMaximumValue,
        maxAttribute,
        expressionMaximumValue,
        noOfMarkers,
        decimalPlaces,
        tooltip,
        showTooltip,
        tooltipType,
        tooltipAlwaysVisible,
        onChange: onChangeProp
    } = props;
    const minValue = getMinValue({
        minValueType,
        staticMinimumValue,
        minAttribute,
        expressionMinimumValue
    });
    const maxValue = getMaxValue({
        maxValueType,
        staticMaximumValue,
        maxAttribute,
        expressionMaximumValue
    });
    const step = getStepValue({
        stepSizeType,
        stepValue,
        stepAttribute,
        expressionStepSize
    });
    const marks = useMarks({
        noOfMarkers,
        decimalPlaces,
        minValueType,
        staticMinimumValue,
        minAttribute,
        expressionMinimumValue,
        maxValueType,
        staticMaximumValue,
        maxAttribute,
        expressionMaximumValue
    });
    const handle = createHandleGenerator({ tooltip, showTooltip, tooltipType, tooltipAlwaysVisible });
    const { onChange } = useOnChangeDebounced({ valueAttribute, onChange: onChangeProp });
    const style = getStyleProp({ orientation, height, heightUnit });
    useScheduleUpdateOnce(() => valueAttribute.status === ValueStatus.Available);

    return (
        <SliderComponent
            disabled={valueAttribute.readOnly}
            rootStyle={style}
            vertical={isVertical(props)}
            value={valueAttribute.value?.toNumber()}
            min={minValue}
            max={maxValue}
            step={step}
            onChange={onChange}
            marks={marks}
            handle={handle}
        />
    );
}
