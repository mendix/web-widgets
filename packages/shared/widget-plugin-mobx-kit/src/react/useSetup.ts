import { useEffect } from "react";
import { useConst } from "./useConst";

export function useSetup<
    T extends {
        setup(): void | (() => void);
    }
>(fn: () => T): T {
    const obj = useConst(fn);
    useEffect(() => obj.setup(), [obj]);
    return obj;
}
