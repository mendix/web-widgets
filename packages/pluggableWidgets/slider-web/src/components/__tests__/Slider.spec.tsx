import { cleanup, fireEvent, render, screen } from "@testing-library/react";

import { Slider, SliderProps } from "../Slider";

describe("Slider", () => {
    afterEach(cleanup);

    const defaultSliderProps: SliderProps = {
        min: -100,
        max: 100,
        step: 10
    };

    function renderSlider(props?: Partial<SliderProps>): ReturnType<typeof render> {
        return render(<Slider {...defaultSliderProps} {...props} />);
    }

    it("renders horizontal Slider correctly", () => {
        const { asFragment } = renderSlider();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders vertical Slider correctly", () => {
        const { asFragment } = renderSlider({ vertical: true });
        expect(asFragment()).toMatchSnapshot();
    });

    it("contains correct value", () => {
        renderSlider({ value: 30 });
        const handle = screen.getByRole("slider");
        expect(handle.getAttribute("aria-valuenow")).toBe("30");
    });

    it("handles keydown events", async () => {
        const onChange = jest.fn();
        render(<Slider {...defaultSliderProps} onChange={onChange} />);
        const sliderHandle = screen.getByRole("slider");
        sliderHandle.focus();

        fireEvent.keyDown(sliderHandle, { key: "ArrowUp", keyCode: 38, bubbles: true });
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange.mock.calls[0][0]).toEqual(-90);

        fireEvent.keyDown(sliderHandle, { key: "ArrowLeft", keyCode: 37, bubbles: true });
        expect(onChange).toHaveBeenCalledTimes(2);
        expect(onChange.mock.calls[1][0]).toEqual(-100);

        fireEvent.keyDown(sliderHandle, { key: "ArrowRight", keyCode: 39, bubbles: true });
        expect(onChange).toHaveBeenCalledTimes(3);
        expect(onChange.mock.calls[2][0]).toEqual(-90);
    });

    it("renders markers correctly", () => {
        const marks = {
            [-100]: "-100",
            [-50]: "-50",
            0: "0",
            50: "50",
            100: "100"
        };
        const sliderProps: SliderProps = {
            ...defaultSliderProps,
            marks
        };

        const { asFragment } = render(<Slider {...sliderProps} />);

        expect(asFragment()).toMatchSnapshot();

        for (const label of Object.values(marks)) {
            screen.getByText(new RegExp(`^${label}$`, "i"));
        }
    });
});
