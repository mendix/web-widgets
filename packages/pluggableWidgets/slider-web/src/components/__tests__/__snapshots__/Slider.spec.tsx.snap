// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders horizontal Slider correctly 1`] = `
<DocumentFragment>
  <div
    class="widget-slider"
  >
    <div
      class="rc-slider rc-slider-horizontal"
    >
      <div
        class="rc-slider-rail"
      />
      <div
        class="rc-slider-track"
        style="left: 0%; width: 0%;"
      />
      <div
        class="rc-slider-step"
      />
      <div
        aria-disabled="false"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="-100"
        aria-valuenow="-100"
        class="rc-slider-handle"
        role="slider"
        style="left: 0%; transform: translateX(-50%);"
        tabindex="0"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Slider renders markers correctly 1`] = `
<DocumentFragment>
  <div
    class="widget-slider"
  >
    <div
      class="rc-slider rc-slider-horizontal rc-slider-with-marks"
    >
      <div
        class="rc-slider-rail"
      />
      <div
        class="rc-slider-track"
        style="left: 0%; width: 0%;"
      />
      <div
        class="rc-slider-step"
      >
        <span
          class="rc-slider-dot rc-slider-dot-active"
          style="left: 0%; transform: translateX(-50%);"
        />
        <span
          class="rc-slider-dot"
          style="left: 25%; transform: translateX(-50%);"
        />
        <span
          class="rc-slider-dot"
          style="left: 50%; transform: translateX(-50%);"
        />
        <span
          class="rc-slider-dot"
          style="left: 75%; transform: translateX(-50%);"
        />
        <span
          class="rc-slider-dot"
          style="left: 100%; transform: translateX(-50%);"
        />
      </div>
      <div
        aria-disabled="false"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="-100"
        aria-valuenow="-100"
        class="rc-slider-handle"
        role="slider"
        style="left: 0%; transform: translateX(-50%);"
        tabindex="0"
      />
      <div
        class="rc-slider-mark"
      >
        <span
          class="rc-slider-mark-text rc-slider-mark-text-active"
          style="left: 0%; transform: translateX(-50%);"
        >
          -100
        </span>
        <span
          class="rc-slider-mark-text"
          style="left: 25%; transform: translateX(-50%);"
        >
          -50
        </span>
        <span
          class="rc-slider-mark-text"
          style="left: 50%; transform: translateX(-50%);"
        >
          0
        </span>
        <span
          class="rc-slider-mark-text"
          style="left: 75%; transform: translateX(-50%);"
        >
          50
        </span>
        <span
          class="rc-slider-mark-text"
          style="left: 100%; transform: translateX(-50%);"
        >
          100
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Slider renders vertical Slider correctly 1`] = `
<DocumentFragment>
  <div
    class="widget-slider widget-slider-vertical"
  >
    <div
      class="rc-slider rc-slider-vertical"
    >
      <div
        class="rc-slider-rail"
      />
      <div
        class="rc-slider-track"
        style="bottom: 0%; height: 0%;"
      />
      <div
        class="rc-slider-step"
      />
      <div
        aria-disabled="false"
        aria-orientation="vertical"
        aria-valuemax="100"
        aria-valuemin="-100"
        aria-valuenow="-100"
        class="rc-slider-handle"
        role="slider"
        style="bottom: 0%; transform: translateY(50%);"
        tabindex="0"
      />
    </div>
  </div>
</DocumentFragment>
`;
